app/models/protocols_participant.rb

```diff
@@ -0,0 +1,57 @@
+class ProtocolsParticipant < ApplicationRecord
+  belongs_to :protocol
+  belongs_to :participant
+  belongs_to :arm
+
+  has_many :appointments
+  has_many :procedures, through: :appointments
+
+  after_save :update_faye
+  after_destroy :update_faye
+
+  validates :protocol_id, presence: true
+  validates :participant_id, presence: true
+
+  def build_appointments
+    ActiveRecord::Base.transaction do
+      if arm
+        if appointments.empty?
+          appointments_for_visit_groups(arm.visit_groups)
+        elsif has_new_visit_groups?
+          appointments_for_visit_groups(new_visit_groups)
+        end
+      end
+    end
+  end
+
+  def update_appointments_on_arm_change
+    appointments.each{ |appt| appt.destroy_if_incomplete }
+    build_appointments
+  end
+
+  def can_be_destroyed?
+    procedures.where.not(status: 'unstarted').empty?
+  end
+
+  private
+
+  def has_new_visit_groups?
+    arm.visit_groups.order(:id).pluck(:id) != appointments.where(arm_id: arm_id).where.not(visit_group_id: nil).order(:visit_group_id).pluck(:visit_group_id)
```

Comment:
Would it make sense to call `self.new_visit_groups.any?` here rather than this query?

Reply:
Created a new story for this optimization:  https://www.pivotaltracker.com/story/show/164875068